eclair-core/src/main/scala/fr/acinq/eclair/payment/send/MultiPartPaymentLifecycle.scala

```diff
@@ -230,13 +230,14 @@ class MultiPartPaymentLifecycle(nodeParams: NodeParams, cfg: SendPaymentConfig,
     event match {
       case Left(paymentFailed) =>
         log.warning("multi-part payment failed")
+        log.info(s"failed payment attempts details:\n    + ${paymentFailed.failures.map(f => PaymentFailure.print(f)).mkString("\n    + ")}")
```

Comment:
I would split the string first, and then `foreach(log.info(...))`. That way all lines are prefixed by the MDC and can be grepped at once.

An alternative to this would be to print a compact (meaning: no line breaks) json representation of an object. That way the formatting would be defined by the object structure, instead of clever indenting. Another advantage is that we can store and parse this debug info in db, and also we can encapsulate mutliple payment attempts in a higher-level object. WDYT?

Reply:
> I would split the string first, and then foreach(log.info(...)). That way all lines are prefixed by the MDC and can be grepped at once.

That's what I actually want to avoid, because then we need to grep (because unrelated log lines may have been inserted), and each individual log line is confusing when taken outside of the context of the others...I really want this to be in a single log, like exceptions stack traces, so that it's really easy to just read in a log file.

> An alternative to this would be to print a compact (meaning: no line breaks) json representation of an object. 

I considered that, and I'm hesitating. I chose against it because my first need is readability / quick understanding of the payment breakdown, similarly to printed stack traces. But a JSON object could easily be copy-pasted in a `.json` file and easy to read, it's one extra step but it's not too bad.

I don't know what to decide, let me know what you prefer if you can help me tilt the decision towards one or the other!

Reply:
> That's what I actually want to avoid, because then we need to grep (because unrelated log lines may have been inserted), and each individual log line is confusing when taken outside of the context of the others...I really want this to be in a single log, like exceptions stack traces, so that it's really easy to just read in a log file.

You mean in a separate log file? But there will still be concurrency between several independent payments and logs will be intertwinned (altough less frequently, granted). I think there is a case for putting the MDC even if we go for a separate file.

> I considered that, and I'm hesitating. I chose against it because my first need is readability / quick understanding of the payment breakdown, similarly to printed stack traces. But a JSON object could easily be copy-pasted in a .json file and easy to read, it's one extra step but it's not too bad.
> I don't know what to decide, let me know what you prefer if you can help me tilt the decision towards one or the other!

Here is another argument in favor of json: we could return it in our API response when some debug flag is enabled.

Reply:
> You mean in a separate log file?

No, I meant a single `log` call to ensure it's atomic and easy to read in one go.
This way you're sure that this will not be intertwined with other, unrelated logs.

> Here is another argument in favor of json: we could return it in our API response when some debug flag is enabled.

Ok let's go for a JSON structure then.

Reply:
Done in https://github.com/ACINQ/eclair/pull/1966/commits/26d4a7699f78eb3a6e41ce1ede21bb704c31f549

Reply:
> I meant a single log call to ensure it's atomic and easy to read in one go.
> This way you're sure that this will not be intertwined with other, unrelated logs.

Oh I see. I thought multi-line logs would be intertwined but maybe you are right!

Reply:
Actually I'm not sure at all, I suppose they're not but it's hard to actually verify!